okhttp/src/main/java/okhttp3/HttpUrl.java

```diff
@@ -13,1331 +13,1173 @@
    */
-  public String redact() {
-    return newBuilder("/...")
+  fun redact(): String {
+    return newBuilder("/...")!!
         .username("")
         .password("")
         .build()
-        .toString();
+        .toString()
   }
 
   /**
-   * Returns the URL that would be retrieved by following {@code link} from this URL, or null if
-   * the resulting URL is not well-formed.
+   * Returns the URL that would be retrieved by following `link` from this URL, or null if the
+   * resulting URL is not well-formed.
    */
-  public @Nullable HttpUrl resolve(String link) {
-    Builder builder = newBuilder(link);
-    return builder != null ? builder.build() : null;
-  }
-
-  public Builder newBuilder() {
-    Builder result = new Builder();
-    result.scheme = scheme;
-    result.encodedUsername = encodedUsername();
-    result.encodedPassword = encodedPassword();
-    result.host = host;
+  fun resolve(link: String): HttpUrl? = newBuilder(link)?.build()
+
+  fun newBuilder(): Builder {
+    val result = Builder()
+    result.scheme = scheme
+    result.encodedUsername = encodedUsername()
+    result.encodedPassword = encodedPassword()
+    result.host = host
     // If we're set to a default port, unset it in case of a scheme change.
-    result.port = port != defaultPort(scheme) ? port : -1;
-    result.encodedPathSegments.clear();
-    result.encodedPathSegments.addAll(encodedPathSegments());
-    result.encodedQuery(encodedQuery());
-    result.encodedFragment = encodedFragment();
-    return result;
+    result.port = if (port != defaultPort(scheme)) port else -1
+    result.encodedPathSegments.clear()
+    result.encodedPathSegments.addAll(encodedPathSegments())
+    result.encodedQuery(encodedQuery())
+    result.encodedFragment = encodedFragment()
+    return result
   }
 
   /**
-   * Returns a builder for the URL that would be retrieved by following {@code link} from this URL,
+   * Returns a builder for the URL that would be retrieved by following `link` from this URL,
    * or null if the resulting URL is not well-formed.
    */
-  public @Nullable Builder newBuilder(String link) {
+  fun newBuilder(link: String): Builder? {
     try {
-      return new Builder().parse(this, link);
-    } catch (IllegalArgumentException ignored) {
-      return null;
+      return Builder().parse(this, link)
+    } catch (ignored: IllegalArgumentException) {
+      return null
     }
   }
 
-  /**
-   * Returns a new {@code HttpUrl} representing {@code url} if it is a well-formed HTTP or HTTPS
-   * URL, or null if it isn't.
-   */
-  public static @Nullable HttpUrl parse(String url) {
-    try {
-      return get(url);
-    } catch (IllegalArgumentException ignored) {
-      return null;
-    }
+  override fun equals(other: Any?): Boolean {
+    return other is HttpUrl && other.url == url
   }
 
-  /**
-   * Returns a new {@code HttpUrl} representing {@code url}.
-   *
-   * @throws IllegalArgumentException If 
```

Comment:
Just read about a neat "non-nullable shadow" pattern, you just do
```kotlin
val encodedQueryNamesAndValues = encodedQueryNamesAndValues ?: throw NullPointerException()
```
and the rest of the code in the method doesn't need the double bangs